numba/inline_closurecall.py

```diff
@@ -203,19 +246,435 @@ def _replace_returns(blocks, target, return_label):
+                debug_print("iter_def = ", iter_def)
+                iter_vars.append(expr.value)
+            elif expr.op == 'pair_first':
+                iter_first_vars.append(stmt.target)
+
+    # Require only one iterator in loop header
+    require(len(iter_vars) == 1 and len(iter_first_vars) == 1)
+    iter_var = iter_vars[0] # variable that holds the iterator object
+    iter_first_var = iter_first_vars[0] # variable that holds the value out of iterator
+
+    # Final requirement: only one loop entry, and we're going to modify it by:
+    # 1. replacing the list definition with an array definition;
+    # 2. adding a counter for the array iteration.
+    require(len(loop.entries) == 1)
+    loop_entry = func_ir.blocks[next(iter(loop.entries))]
+    terminator = loop_entry.terminator
+    scope = loop_entry.scope
+    loc = loop_entry.loc
+    stmts = []
+    removed = []
+    def is_removed(val, removed):
+        if isinstance(val, ir.Var):
+            for x in removed:
+                if x.name == val.name:
+                    return True
+        return False
+    # Skip list construction and skip terminator, add the rest to stmts
+    for i in range(len(loop_entry.body) - 1):
+        stmt = loop_entry.body[i]
+        if isinstance(stmt, ir.Assign) and (stmt.value == list_def or is_removed(stmt.value, removed)):
+            removed.append(stmt.target)
+        else:
+            stmts.append(stmt)
+    debug_print("removed variables: ", removed)
+
+    # Define an index_var to index the array.
+    # If the range happens to be single step ranges like range(n), or range(m, n),
+    # then the index_var correlates to iterator index; otherwise we'll have to
+    # define a new counter.
+    range_def = guard(_find_iter_range, func_ir, iter_var)
+    index_var = scope.make_temp(loc)
+    index_offset_var = None
+    if range_def and range_def[0] == 0:
+        # iterator starts with 0, index_var can just be iter_first_var
+        index_var = iter_first_var
+    else:
+        # index_var = -1 # starting the index with -1 since it will incremented in loop header
+        stmts.append(_new_definition(func_ir, index_var, ir.Const(value=-1, loc=loc), loc))
+
+    # Insert statement to get the size of the loop iterator
+    size_var = scope.make_temp(loc)
+    if range_def:
+        start, stop = range_def
+        # simple range with start and stop, which are defined only in caller,
+        # So we'll have to pass them as parameters
+        if start == 0:
+            # size_val = stop
+            size_val = stop
+        else:
+            # size_val = stop - start
+            size_val = ir.Expr.binop(fn='-', lhs=stop, rhs=start, loc=loc)
+    else:
+        len_func_var = scope.make_temp(loc)
+        stmts.append(_new_definition(func_ir, len_func_var,
+                     ir.Global('range_iter_len', range_iter_len, loc=loc), loc))
+        # size_var = len_func(iter_var)
```

Comment:
will remove the commented line since the next line is pretty much self-explanatory.